### PR TITLE
✨ Conditions for OpenStackMachines

### LINKED ISSUE #832: Implement conditions
/kind feature

**Describe the solution you'd like**

Cluster API has [conditions](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200506-conditions.md#kubernetes-conditions) feature. CAPI/CAPA/CAPZ/CAPV already implemented the feature. we should have it, too.

**Anything else you would like to add:**

### PR DESCRIPTION
**Pull request by apricote**
**What this PR does / why we need it**:

Adds Conditions to the OpenStackMachineStatus.

See this example `clusterctl describe` output, that showcases the different conditions, and how they "bubble up" to the `Machine` resources:
screenshot-capo-machine-conditions

So far there are two conditions: 
- `InstanceReady` (related to the OpenStack Instance) and 
- `APIServerIngressReadyCondition` (for ControlPlane machines -> LoadBalancer Member or Floating IP attachment)

I set the severity based on:

- *Info*, if the reason is expected (missing bootstrap secret)
- *Warning*, if the condition is eventually retried (no call to `handleUpdateMachineError`)
- *Error* if `handleUpdateMachineError` was called and the machine is no longer retried/reconciled

The `Ready` condition is set to the aggregate of thhe two custom conditions, and the CAPI Machine resource "bubbles up" to Ready message, so any errors can be observed easily, without having to inspect the OpenStackMachine

**Which issue(s) this PR fixes**:
Fixes #832 
Replaces #972

**Open Tasks**:

- [ ] ~~Tests for OpenStackMachineController~~ Requires proper envtest mocks
- [x] Fix `APIServerIngressReadyCondition` on non-control-plane machines when using api loadbalancer (see #1294)

**Open Questions**:

*Should we add the conditions to all existing api versions (because its a backwards compatible change) or only to the most recent version (v1alpha6)?*

  - **Pro**: Makes conversion a no-op
  - **Pro**: Changes are backwards compatible
  - **Contra**: Changing old API versions feels wrong to me

/hold

### CODE REVIEW

File: controllers/openstackmachine_controller.go [hunk 1]
```diff
@@ -242,6 +270,7 @@ func (r *OpenStackMachineReconciler) reconcileDelete(ctx context.Context, scope
 				if address.Type == corev1.NodeExternalIP {
 					if err = networkingService.DeleteFloatingIP(openStackMachine, address.Address); err != nil {
 						handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("error deleting Openstack floating IP: %v", err))
+						conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.FloatingIPErrorReason, clusterv1.ConditionSeverityError, "error deleting floating IP: %v", err)
```
Comment (tobiasgiese) [inline controllers/openstackmachine_controller.go:273]:
```suggestion
						conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.FloatingIPErrorReason, clusterv1.ConditionSeverityError, "Deleting floating IP failed: %v", err)
```

File: controllers/openstackmachine_controller.go [hunk 2]
```diff
@@ -253,11 +282,13 @@ func (r *OpenStackMachineReconciler) reconcileDelete(ctx context.Context, scope
 	if err != nil {
 		err = errors.Errorf("machine spec is invalid: %v", err)
 		handleUpdateMachineError(scope.Logger, openStackMachine, err)
+		conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InvalidMachineSpecReason, clusterv1.ConditionSeverityError, err.Error())
 		return ctrl.Result{}, err
 	}
 
 	if err := computeService.DeleteInstance(openStackMachine, instanceSpec, instanceStatus); err != nil {
 		handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("error deleting OpenStack instance %s with ID %s: %v", instanceStatus.Name(), instanceStatus.ID(), err))
+		conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InstanceDeleteFailedReason, clusterv1.ConditionSeverityError, "error deleting instance: %v", err)
```
Comment (tobiasgiese) [inline controllers/openstackmachine_controller.go:291]:
```suggestion
		conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InstanceDeleteFailedReason, clusterv1.ConditionSeverityError, "Deleting instance failed: %v", err)
```

File: controllers/openstackmachine_controller.go [hunk 3]
```diff
@@ -373,19 +415,24 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
 		fp, err := networkingService.GetOrCreateFloatingIP(openStackMachine, openStackCluster, clusterName, floatingIPAddress)
 		if err != nil {
 			handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("Floating IP cannot be got or created: %v", err))
+			conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.FloatingIPErrorReason, clusterv1.ConditionSeverityError, "Floating IP can not be got or created: %v", err)
```
Comment (tobiasgiese) [inline controllers/openstackmachine_controller.go:418]:
```suggestion
			conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.FloatingIPErrorReason, clusterv1.ConditionSeverityError, "Floating IP cannot be obtained or created: %v", err)
```

**Review by tobiasgiese — COMMENTED** [review]
Implementing conditions is a really huge improvement, tbh! I like it.
Some nits, but @mdbooth can certainly help better here 🙂

**Comment by seanschneeweiss** [comment]
This PR is such a great addon for CAPO. Thanks a lot.

> Should we add the conditions to all existing api versions (because its a backwards compatible change) or only to the most recent version (v1alpha6)?

I think it is totally fine to add it to the recent version only.

**Comment by apricote** [comment]
I rebased the PR against the current `main` commit. The bug where the `APIServerIngressReady` condition was added to non-CP machines is fixed now.

The PR is ready for review.

**Comment by apricote** [comment]
> ERROR: (gcloud.compute.instances.create) Could not fetch resource:
> - The zone 'projects/k8s-boskos-gce-project-12/zones/us-east4-a' does not have enough resources available to fulfill the request.  '(resource type:compute)'. 

/retest

**Review by seanschneeweiss — COMMENTED** [review]
/lgtm
:+1:

File: controllers/openstackmachine_controller.go [hunk 4]
```diff
@@ -368,6 +407,7 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
 		err = r.reconcileLoadBalancerMember(scope, openStackCluster, machine, openStackMachine, instanceNS, clusterName)
 		if err != nil {
 			handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("LoadBalancerMember cannot be reconciled: %v", err))
+			conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.LoadBalancerMemberErrorReason, clusterv1.ConditionSeverityError, "Reconciling LoadBalancerMember failed: %v", err)
```
Comment (tobiasgiese) [inline controllers/openstackmachine_controller.go:410]:
```suggestion
			conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.LoadBalancerMemberErrorReason, clusterv1.ConditionSeverityError, "Reconciling load balancer member failed: %v", err)
```

I think it's better to write this out in the condition. The condition is for the end user as well, not only for us devs :)

File: controllers/openstackmachine_controller.go [hunk 5]
```diff
@@ -393,10 +435,12 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
 			err = networkingService.AssociateFloatingIP(openStackMachine, fp, port.ID)
 			if err != nil {
 				handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("Floating IP cannot be associated: %v", err))
+				conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.FloatingIPErrorReason, clusterv1.ConditionSeverityError, "Associating floating IP  failed: %v", err)
```
Comment (tobiasgiese) [inline controllers/openstackmachine_controller.go:438]:
```suggestion
				conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.FloatingIPErrorReason, clusterv1.ConditionSeverityError, "Associating floating IP failed: %v", err)
```

Sorry for nit-picking 😅

**Review by tobiasgiese — APPROVED** [review]
All in all this looks pretty solid 👌🏻 
/approve

**Comment by apricote** [comment]
Tests will be fixed by #1313. I will rebase once that PR is merged.

File: controllers/openstackmachine_controller.go [hunk 6]
```diff
@@ -414,11 +458,13 @@ func (r *OpenStackMachineReconciler) getOrCreate(logger logr.Logger, cluster *cl
 		if err != nil {
 			err = errors.Errorf("machine spec is invalid: %v", err)
 			handleUpdateMachineError(logger, openStackMachine, err)
+			conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InvalidMachineSpecReason, clusterv1.ConditionSeverityError, err.Error())
 			return nil, err
 		}
 
 		instanceStatus, err = computeService.CreateInstance(openStackMachine, openStackCluster, instanceSpec, cluster.Name)
 		if err != nil {
+			conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InstanceCreateFailedReason, clusterv1.ConditionSeverityError, err.Error())
```
Comment (seanschneeweiss) [inline controllers/openstackmachine_controller.go:467]: Condition severity Warning might be better here, as this is very likely to be a retryable error.

File: controllers/openstackmachine_controller.go [hunk 7]
```diff
@@ -393,10 +435,12 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
 			err = networkingService.AssociateFloatingIP(openStackMachine, fp, port.ID)
 			if err != nil {
 				handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("Floating IP cannot be associated: %v", err))
+				conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.FloatingIPErrorReason, clusterv1.ConditionSeverityError, "Associating floating IP failed: %v", err)
```
Comment (seanschneeweiss) [inline controllers/openstackmachine_controller.go:438]: Condition severity Warning might be better here, as this is very likely to be a retryable error.

File: controllers/openstackmachine_controller.go [hunk 8]
```diff
@@ -378,12 +418,14 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
...
 		if err != nil {
 			handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("Floating IP cannot be got or created: %v", err))
+			conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.FloatingIPErrorReason, clusterv1.ConditionSeverityError, "Floating IP cannot be obtained or created: %v", err)
 			return ctrl.Result{}, nil
 		}
 		port, err := computeService.GetManagementPort(openStackCluster, instanceStatus)
 		if err != nil {
 			err = errors.Errorf("getting management port for control plane machine %s: %v", machine.Name, err)
 			handleUpdateMachineError(scope.Logger, openStackMachine, err)
+			conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.FloatingIPErrorReason, clusterv1.ConditionSeverityError, "Obtaining management port for control plane machine failed: %v", err)
```
Comment (seanschneeweiss) [inline controllers/openstackmachine_controller.go:428]: Condition severity Warning might be better here, as this is very likely to be a retryable error.

File: controllers/openstackmachine_controller.go [hunk 9]
```diff
@@ -368,6 +407,7 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
 		err = r.reconcileLoadBalancerMember(scope, openStackCluster, machine, openStackMachine, instanceNS, clusterName)
 		if err != nil {
 			handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("LoadBalancerMember cannot be reconciled: %v", err))
+			conditions.MarkFalse(openStackMachine, infrav1.APIServerIngressReadyCondition, infrav1.LoadBalancerMemberErrorReason, clusterv1.ConditionSeverityError, "Reconciling load balancer member failed: %v", err)
```
Comment (seanschneeweiss) [inline controllers/openstackmachine_controller.go:410]: Condition severity Warning might be better here, as this is very likely to be a retryable error.

File: controllers/openstackmachine_controller.go [hunk 10]
```diff
@@ -343,19 +378,23 @@ func (r *OpenStackMachineReconciler) reconcileNormal(ctx context.Context, scope
...
 		openStackMachine.Status.Ready = true
 	case infrav1.InstanceStateError:
 		// Error is unexpected, thus we report error and never retry
 		handleUpdateMachineError(scope.Logger, openStackMachine, errors.Errorf("OpenStack instance state %q is unexpected", instanceStatus.State()))
+		conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InstanceStateErrorReason, clusterv1.ConditionSeverityError, "")
 		return ctrl.Result{}, nil
 	case infrav1.InstanceStateDeleted:
 		// we should avoid further actions for DELETED VM
 		scope.Logger.Info("Instance state is DELETED, no actions")
+		conditions.MarkFalse(openStackMachine, infrav1.InstanceReadyCondition, infrav1.InstanceDeletedReason, clusterv1.ConditionSeverityError, "")
```
Comment (seanschneeweiss) [inline controllers/openstackmachine_controller.go:391]: This might be an expected state, so don't know if we should really considers this error severity.

**Review by seanschneeweiss — COMMENTED** [review]
I wonder if the severity of some conditions should be considered as Warning and not as Error. WDYT?

**Comment by seanschneeweiss** [comment]
> I wonder if the severity of some conditions should be considered as Warning and not as Error. WDYT?

Oh sorry, I guess the severity can be adapted as soon as the handleUpdateMachineError is removed in #1116.

**Review by seanschneeweiss — APPROVED** [review]

**Pull request merged**